### PR TITLE
Add support for an extra set of kernel packages and group configured with NO_HZ_FULL

### PR DESCRIPTION
**Pull request by gratian**
In certain types of real-time applications it is desirable to reserve CPU cores for real-time threads and avoid interference from the kernel as much as possible. This includes things like the kernel scheduler tick which can create jitter spikes.

Provide packages for a kernel configured with NO_HZ_FULL. This allows users with strict RT requirements to install a kernel that will adaptively try to shutdown the scheduler thick whenever possible. This comes at the expense of some overhead in user <-> kernel transitions (i.e. syscalls, exceptions, and interrupts).

This necessitaded some refactoring of existing recipes which assumed only two possible kernel configurations: 'current' and 'next'

### CODE REVIEW

**Comment by gratian** [comment]
Attn: @ni/rtos 
FYI: @jhershbe

**Comment by jhershbe** [comment]
Looks good to me.

**Review by amstewart — CHANGES_REQUESTED** [review]
`linux-nilrt-extra` kind of sounds like it installs non-essential kernel components. What about `linux-nilrt-alternate` instead?

File: recipes-kernel/linux/linux-nilrt-next_git.bb [hunk 1]
```diff
@@ -1,6 +1,6 @@
 DESCRIPTION = "NILRT linux kernel next development build"
 
-require linux-nilrt-next.inc
+require linux-nilrt-extra.inc
```
Comment (amstewart) [inline recipes-kernel/linux/linux-nilrt-next_git.bb]: linux-nilrt-extra kind of sounds like it installs non-essential kernel components. What about linux-nilrt-alternate instead?

**Review by amstewart — CHANGES_REQUESTED** [review]
Disregard the above comment. Let's try this again, github.

File: recipes-kernel/linux/linux-nilrt.inc [hunk 2]
```diff
@@ -1,10 +1,18 @@
...
 machine_srcrev="${SRCREV}"
 
+# Subfolder of the same name will be added to FILESEXTRAPATHS and also
+# used for nilrt-specific config fragment manipulation during build.
+# Provide a unique name for each recipe saved in the same source folder
+KBUILD_FRAGMENTS_LOCATION ?= "nilrt"
+
 # Set the version for traceability purposes
-LINUX_VERSION_EXTENSION = "$([ -z "${BUILDNAME}" ] || echo '-')${BUILDNAME}"
+LINUX_VERSION_EXTENSION = "$([ "${LINUX_KERNEL_TYPE}" != "standard" ] && echo '-${LINUX_KERNEL_TYPE}')$([ -z "${BUILDNAME}" ] || echo '-')${BUILDNAME}"
```
Comment (amstewart) [inline recipes-kernel/linux/linux-nilrt.inc:15]: Does the `LINUX_KERNEL_TYPE` get set to `standard` in the "current" kernel recipe? I don't see that happening anywhere.

File: recipes-kernel/packagegroups/packagegroup-ni-debug-kernel.bb [hunk 3]
```diff
@@ -5,12 +5,11 @@ inherit packagegroup
 
 RDEPENDS_${PN} = "\
 	kernel-debug \
-	kernel-debug-base \
-	kernel-debug-vmlinux \
-	kernel-debug-image \
```
Comment (amstewart) [inline recipes-kernel/packagegroups/packagegroup-ni-debug-kernel.bb]: Why are we dropping all these from the packagegroup? Are they deprecated?

File: recipes-kernel/linux/linux-nilrt-next_git.bb (hunk 1 above)
Comment (gratian) [reply recipes-kernel/linux/linux-nilrt-next_git.bb]: Sounds good. I'll  change the name on the next patch rev.

File: recipes-kernel/linux/linux-nilrt.inc (hunk 2 above)
Comment (gratian) [reply recipes-kernel/linux/linux-nilrt.inc:15]: meta-nilrt/recipes-kernel/linux/linux-nilrt.inc (which is used by all nilrt kernel recipes) includes openembedded-core/meta/recipes-kernel/linux/linux-yocto.inc which sets:

```
# The distro or local.conf should set this, but if nobody cares...
LINUX_KERNEL_TYPE ??= "standard"
```
We've been using this default for virtual/kernel package builds. There might be an argument to be made here about setting the (weak defined) default in distro or local.conf as the comment in that .inc file implies but we've been operating using the linux-yocto.inc default for a while so I didn't mess with it.

Comment (amstewart) [reply recipes-kernel/linux/linux-nilrt.inc:15]: OK. That's sufficient, I think.

File: recipes-kernel/packagegroups/packagegroup-ni-debug-kernel.bb (hunk 3 above)
Comment (gratian) [reply recipes-kernel/packagegroups/packagegroup-ni-debug-kernel.bb]: The packagegroup is over-specifying the dependencies. The dependency chain for _kernel-debug_ already includes the _kernel-debug-base_ dependency which in turn depends on the package providing _kernel-debug-image_. Here's the dependency chain reconstructed from .ipks I have built to test this:

```
  kernel-debug_4.14+git0+db94e1d70a-r0_x64.ipk
  Depends: kernel-debug-4.14.146-rt67-debug-dev-nilrt

	kernel-debug-4.14.146-rt67-debug-dev-nilrt_4.14+git0+db94e1d70a-r0_x64.ipk
	Depends: kernel-debug-image-4.14.146-rt67-debug-dev-nilrt
	Provides: kernel-debug-4.14.146-rt67-debug-dev-nilrt, kernel-debug-base

		kernel-debug-image-4.14.146-rt67-debug-dev-nilrt_4.14+git0+db94e1d70a-r0_x64.ipk
		Depends: kernel-debug-image-bzimage-4.14.146-rt67-debug-dev-nilrt
		Provides: kernel-debug-image

			kernel-debug-image-bzimage-4.14.146-rt67-debug-dev-nilrt_4.14+git0+db94e1d70a-r0_x64.ipk
			Provides: kernel-debug-image-bzimage

```
The removal of _kernel-debug-vmlinux_ is accidental and caused by the fact I was reusing the nohz/next package groups (which don't need/want the vmlinux binary). For the debug kernel the vmlinux package is needed bc. it contains the un-stripped kernel image with debug symbols. Thanks for catching this; will fix in V2.

**Comment by gratian** [comment]
Changes since V1:

  * rebase on latest nilrt/master/sumo

  * rename linux-nilrt-extra.inc -> linux-nilrt-alternate.inc and update corresponding recipes

  * re-add kernel-debug-vmlinux to packagegroup-ni-debug-kernel.bb

**Review by amstewart — APPROVED** [review]

**Comment by amstewart** [comment]
@chaitu236 Any objections?

**Review by chaitu236 — APPROVED** [review]

**Comment by chaitu236** [comment]
> @chaitu236 Any objections?

No looks good to me.

**Pull request merged**